feat: Add base for the route diagram

mmelko opened the pull request:
This is a WIP for the route diagram based on reactflow+dagre.
What's missing: 
 - fixed styling, currently styles are all over the place, that needs to be seriously polished 
 - attributes for each node 
 
 current screenshot: 
<img width="623" alt="Screenshot 2023-03-29 at 16 33 39" src="https://user-images.githubusercontent.com/6814482/228573198-b7940063-c730-40c7-a670-7cd07bf82c0e.png">

tadayosi commented on packages/hawtio/src/plugins/camel/routes/RouteDiagram.tsx:
Actually, Routes and Route Diagram are two different things in terms of feature. I don't think they share a lot. What about creating another directory `route-diagram` and putting the related code under it instead of putting them under `routes`?

mmelko replied:
Sure, NP .. I've put it in the same directory because it's all in the page under routes but I will move it.

tadayosi commented on packages/hawtio/src/plugins/camel/routes/RouteDiagram.tsx:
```diff
@@ -0,0 +1,127 @@
+const nodeWidth = 300
+const nodeHeight = 86
+
+export type ICamelNode = Node & {
```
You don't need `I` prefix. Non only because it's not an interface, but also because we generally wouldn't use `I` prefixes other than service interfaces.

tadayosi commented on packages/hawtio/src/plugins/camel/routes/RouteDiagram.tsx:
```diff
@@ -0,0 +1,127 @@
+
+  return { nodes, edges }
+}
+const CamelNode = ({ data }: NodeProps<CamelNodeData>) => {
```
For the top-level React components, I'd like to declare type `React.FunctionComponent` so that readers can easily find they are React components .

tadayosi commented on packages/hawtio/src/plugins/camel/routes/visualization-service.ts:
```diff
@@ -0,0 +1,220 @@
+    return siblingNodes
+  }
+}
+export const routeVizualisationService = new VisualizationService()
```
`routeVizualisationService` -> `routeVisualizationService` :-)
Or you can maybe simply name `visualizationService`?

tadayosi commented on packages/hawtio/src/plugins/camel/routes/RouteDiagram.tsx:
```diff
@@ -0,0 +1,127 @@
+  data: CamelNodeData
+}
+
+const getLayoutedElements = (nodes: ICamelNode[], edges: Edge[], direction = 'TB') => {
```
Maybe you can put those dagre specific code, i.e. this function and const `dagreGraph`, into another file or move them into `visualizationService` if that makes sense, in order to keep the React components simple.

tadayosi commented on packages/hawtio/src/plugins/camel/routes/visualization-service.ts:
Is it hard to write tests for the service?

mmelko replied:
I don't think so, I'm planning to add some tests for sure.

phantomjinx commented on packages/hawtio/src/plugins/camel/CamelContent.tsx:
```diff
@@ -78,6 +79,15 @@ export const CamelContent: React.FunctionComponent = () => {
+      title: 'Route Diagram',
+      component: RouteDiagram,
+      isApplicable: (node: MBeanNode) =>
+        !ccs.isEndpointNode(node) &&
```
Checking it is not an Endpoint Node or Folder is probably now unnecessary given they are mutually exclusive to Route Node and Folder.

mmelko replied:
I deleted that check.

phantomjinx reviewed (commented):
LGTM so far

mmelko commented:
@tadayosi I think I fixed all of the issues, also updated a style to look more representative (Still some issues though)

<img width="2114" alt="Screenshot 2023-03-31 at 16 12 10" src="https://user-images.githubusercontent.com/6814482/229144814-77006419-d472-457e-8416-a8c98647990f.png">
I also added simple tests that validates parsing from the service. I think we can merge it in this state if you agree. 

Next I will implement route statistics (on hover and in the right top corner).

tadayosi commented on packages/hawtio/src/plugins/camel/route-diagram/RouteDiagram.tsx:
```diff
@@ -0,0 +1,89 @@
+
+// export type ICamelNode = Node & {
+//   data: CamelNodeData
+// }
```
Can we remove this comment?
